File: packages/schemas/src/domain/ratios.ts
export type RatioGroupKey =
  | 'margins'
  | 'leverage'
  | 'liquidity'
  | 'efficiency'
  | 'valuation_support';

export type RatioDefinition = {
  label: string;
  formula: string;
  note: string;
};

export const RATIO_DEFINITIONS: Record<RatioGroupKey, Record<string, RatioDefinition>> = {
  margins: {
    gross_margin: {
      label: 'Gross Margin',
      formula: 'gross_profit / revenue',
      note: 'Profitability before operating expense burden.',
    },
    operating_margin: {
      label: 'Operating Margin',
      formula: 'operating_income / revenue',
      note: 'Operating profitability per revenue dollar.',
    },
    net_margin: {
      label: 'Net Margin',
      formula: 'net_income / revenue',
      note: 'Bottom-line profitability per revenue dollar.',
    },
    fcf_margin: {
      label: 'Free Cash Flow Margin',
      formula: '(cfo - capex) / revenue',
      note: 'Cash generation after reinvestment intensity.',
    },
  },
  leverage: {
    liabilities_to_equity: {
      label: 'Liabilities to Equity',
      formula: 'liabilities / equity',
      note: 'Simple balance-sheet leverage gauge.',
    },
    liabilities_to_assets: {
      label: 'Liabilities to Assets',
      formula: 'liabilities / assets',
      note: 'Share of assets funded by liabilities.',
    },
  },
  liquidity: {
    current_ratio: {
      label: 'Current Ratio',
      formula: 'current_assets / current_liabilities',
      note: 'Near-term obligations coverage from current assets.',
    },
    quick_ratio: {
      label: 'Quick Ratio',
      formula: '(current_assets - inventory) / current_liabilities',
      note: 'Stricter liquidity excluding inventory.',
    },
  },
  efficiency: {
    asset_turnover: {
      label: 'Asset Turnover',
      formula: 'revenue / assets',
      note: 'Revenue productivity of total assets.',
    },
    receivables_turnover: {
      label: 'Receivables Turnover',
      formula: 'revenue / receivables',
      note: 'Collection efficiency proxy.',
    },
  },
  valuation_support: {
    free_cash_flow: {
      label: 'Free Cash Flow',
      formula: 'cfo - capex',
      note: 'Valuation-ready support metric, not a ratio.',
    },
    eps: {
      label: 'Earnings Per Share',
      formula: 'net_income / shares',
      note: 'Per-share earnings support metric.',
    },
    ev_to_revenue: {
      label: 'EV to Revenue',
      formula: 'enterprise_value / revenue',
      note: 'Simple valuation multiple support metric.',
    },
    price_to_earnings: {
      label: 'Price to Earnings',
      formula: 'market_cap / net_income',
      note: 'Simple equity valuation multiple support metric.',
    },
    fcf_yield: {
      label: 'Free Cash Flow Yield',
      formula: '(cfo - capex) / market_cap',
      note: 'Cash return relative to equity valuation.',
    },
  },
};

File: services/parse-xbrl/src/ratios.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any

FORMULAS: dict[str, dict[str, str]] = {
    "margins": {
        "gross_margin": "gross_profit / revenue",
        "operating_margin": "operating_income / revenue",
        "net_margin": "net_income / revenue",
        "fcf_margin": "(cfo - capex) / revenue",
    },
    "leverage": {
        "liabilities_to_equity": "liabilities / equity",
        "liabilities_to_assets": "liabilities / assets",
    },
    "liquidity": {
        "current_ratio": "current_assets / current_liabilities",
        "quick_ratio": "(current_assets - inventory) / current_liabilities",
    },
    "efficiency": {
        "asset_turnover": "revenue / assets",
        "receivables_turnover": "revenue / receivables",
    },
    "valuation_support": {
        "free_cash_flow": "cfo - capex",
        "eps": "net_income / shares",
        "ev_to_revenue": "enterprise_value / revenue",
        "price_to_earnings": "market_cap / net_income",
        "fcf_yield": "(cfo - capex) / market_cap",
    },
}


@dataclass(frozen=True)
class RatioInputs:
    revenue: float
    gross_profit: float
    operating_income: float
    net_income: float
    assets: float
    liabilities: float
    equity: float
    cfo: float
    capex: float
    current_assets: float
    current_liabilities: float
    inventory: float | None = None
    receivables: float | None = None
    enterprise_value: float | None = None
    market_cap: float | None = None
    shares: float | None = None


def compute_ratio_groups(*, inputs: RatioInputs) -> dict[str, dict[str, float | None]]:
    """Compute explicit, testable ratio groups for deterministic outputs."""

    fcf = inputs.cfo - inputs.capex

    margins = {
        "gross_margin": _safe_divide(inputs.gross_profit, inputs.revenue),
        "operating_margin": _safe_divide(inputs.operating_income, inputs.revenue),
        "net_margin": _safe_divide(inputs.net_income, inputs.revenue),
        "fcf_margin": _safe_divide(fcf, inputs.revenue),
    }

    leverage = {
        "liabilities_to_equity": _safe_divide(inputs.liabilities, inputs.equity),
        "liabilities_to_assets": _safe_divide(inputs.liabilities, inputs.assets),
    }

    inventory = inputs.inventory if inputs.inventory is not None else 0.0
    liquidity = {
        "current_ratio": _safe_divide(inputs.current_assets, inputs.current_liabilities),
        "quick_ratio": _safe_divide(inputs.current_assets - inventory, inputs.current_liabilities),
    }

    efficiency = {
        "asset_turnover": _safe_divide(inputs.revenue, inputs.assets),
        "receivables_turnover": _safe_divide(inputs.revenue, inputs.receivables),
    }

    valuation_support = {
        "free_cash_flow": fcf,
        "eps": _safe_divide(inputs.net_income, inputs.shares),
        "ev_to_revenue": _safe_divide(inputs.enterprise_value, inputs.revenue),
        "price_to_earnings": _safe_divide(inputs.market_cap, inputs.net_income),
        "fcf_yield": _safe_divide(fcf, inputs.market_cap),
    }

    return {
        "margins": margins,
        "leverage": leverage,
        "liquidity": liquidity,
        "efficiency": efficiency,
        "valuation_support": valuation_support,
    }


def formula_catalog() -> dict[str, dict[str, str]]:
    return FORMULAS


def _safe_divide(numerator: float | None, denominator: float | None) -> float | None:
    if numerator is None or denominator is None or denominator == 0:
        return None
    return numerator / denominator


def ratio_snapshot(*, inputs: RatioInputs) -> dict[str, Any]:
    return {
        "inputs": inputs,
        "formulas": formula_catalog(),
        "ratios": compute_ratio_groups(inputs=inputs),
    }
